### ISSUE TITLE
Better Phi for non 2x2 tables

### ISSUE THREAD

**Issue by mattansb** [easystats/effectsize#437]
The general formula for *phi* is:

![image](https://user-images.githubusercontent.com/35330040/165250901-1967e434-5798-47b6-a6ec-b6c34e42e59c.png)

For 2x2 table, this value is bounded between 0 and 1.

## Larger tables

For larger tables, this statistic can be larger than 1, in which case is it advisable to use Cramer's *V* instead.

## 1D table

(aka goodness-of-fit tests)

For 1D tables, phi can also be larger than 1. 

When custom (non-uniform) expected frequencies are used, [Rosenberg (2010)](https://doi.org/10.1371/journal.pone.0010059) proposes to compute *phi* as:

![image](https://user-images.githubusercontent.com/35330040/165250985-0cba4ea7-039c-46fc-87b1-25539890664a.png)

Where ![image](https://user-images.githubusercontent.com/35330040/165251425-c6f37bc0-3fe3-4703-bd2c-dea87b7e4c9d.png) is computed as such:

Given vector `p` (of expected frequency):

```R
p_for_max <- p
p_for_max[ which.min(p)] <- 1
p_for_max[-which.min(p)] <- 0
chisq_max <- effectsize:::.chisq(p_for_max, p)
```

(This is a generalization of the 1x2 case given in Rosenberg's paper)

When the expected frequencies *are* uniform, but there are more than 2 classes, Phi can also be larger than 1 - but I'm not sure how to apply Rosenberg's correction then.

# Action Items:

- [x] Add warning / message when trying to compute *phi* for larger tables?
- [x] Use Rosenberg's method for 1D table? Or are people expecting the inflated value?
  - [x] Can Rosenberg's method for 1D table be applied for the bias-adjusted *phi*?
- [x] Use Rosenberg's method with >1x2 tables with uniform expected frequencies?

Tagging @strengejacke @bwiernik @IndrajeetPatil for your (expert!) opinion - if you have any thoughts...

**Comment by mattansb** [comment]
1D tables:

``` r
library(effectsize)

x <- as.table(matrix(c(100, 30)))

p <- c(0.3, 0.7)

# Manually:
nk <- sum(x) * (p[2]/p[1])
sqrt(chisq.test(x, p = p)$statistic / nk)
#> X-squared 
#> 0.6703297

phi(x, p = p)
#> Phi  |       95% CI
#> -------------------
#> 0.67 | [0.58, 1.00]
#> 
#> - Adjusted Phi for goodness-of-fit.
#> - One-sided CIs: upper bound fixed at [1.00].

phi(x)
#> Phi  |      95% CI
#> ------------------
#> 0.54 | [0.39, Inf]
#> 
#> - One-sided CIs: upper bound fixed at [Inf].



# Larger 1D tables:
x <- as.table(matrix(c(100, 500, 200, 30)))

p <- c(0.7, 0.2, 0.05, 0.05)


Xsq <- chisq.test(x, p = p)


chisq_to_phi(Xsq$statistic, n = sum(Xsq$observed), nrow = 1, ncol = 4)
#> Phi  |      95% CI
#> ------------------
#> 1.42 | [1.36, Inf]
#> 
#> - One-sided CIs: upper bound fixed at [Inf].

chisq_to_phi(Xsq$statistic, n = sum(Xsq$observed), nrow = 1, ncol = 4, 
             p = p)
#> Phi  |       95% CI
#> -------------------
#> 0.33 | [0.31, 1.00]
#> 
#> - Adjusted Phi for goodness-of-fit.
#> - One-sided CIs: upper bound fixed at [1.00].

phi(x, p = p)
#> Phi  |       95% CI
#> -------------------
#> 0.33 | [0.31, 1.00]
#> 
#> - Adjusted Phi for goodness-of-fit.
#> - One-sided CIs: upper bound fixed at [1.00].
```

Warning:

``` r
M <-
  matrix(c(150, 100, 165,
           130, 50, 65,
           35, 10, 2,
           55, 40, 25), nrow = 4,
         dimnames = list(
           Music = c("Pop", "Rock", "Jazz", "Classic"),
           Study = c("Psych", "Econ", "Law")))
M
#>          Study
#> Music     Psych Econ Law
#>   Pop       150   50   2
#>   Rock      100   65  55
#>   Jazz      165   35  40
#>   Classic   130   10  25
phi(M)
#> Warning: For contingency table larger than 2x2, it is advisable to intead
#> estimate Cramer's V.
#> Phi  |       95% CI
#> -------------------
#> 0.34 | [0.27, 1.41]
#> 
#> - One-sided CIs: upper bound fixed at [1.41~].
cramers_v(M)
#> Cramer's V |       95% CI
#> -------------------------
#> 0.24       | [0.19, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].
```

<sup>Created on 2022-04-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by mattansb** [comment]
Seems like the correction works for uniform 1D xtabs. 
Should I apply this? Should this be optional?

**Comment by mattansb** [comment]
``` r
library(effectsize)

x <- as.table(matrix(c(100, 500, 200, 30)))

# uniform expeted
phi(x) 
#> Phi  |       95% CI
#> -------------------
#> 0.50 | [0.47, 1.00]
#> 
#> - Adjusted Phi for goodness-of-fit.
#> - One-sided CIs: upper bound fixed at [1.00].

# custom expeted
phi(x, p = c(0.7, 0.2, 0.05, 0.05)) 
#> Phi  |       95% CI
#> -------------------
#> 0.33 | [0.31, 1.00]
#> 
#> - Adjusted Phi for goodness-of-fit.
#> - One-sided CIs: upper bound fixed at [1.00].
```

<sup>Created on 2022-04-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by bwiernik** [comment]
A phi coefficient is by definition an association between two variables. It doesn't make sense to try to compute it for a 1D chi square (what does that correlation mean)? 

It is also by definition only applicable to a 2x2 (1 df) test statistic. 

The function should error or return NA if anything a 2x2 table chi square (1 df) is provided.

**Comment by strengejacke** [comment]
Based on my limited knowledge, from what I've read and implemented in [sjstats](https://github.com/strengejacke/sjstats/blob/master/R/xtab_statistics.R), is in line what Brenton says. Phi only 2x2, else Cramer.

**Comment by mattansb** [comment]
The general formula for Phi and Cramer's V both normalize the Chi-sq statistic in someway that result in a effect-size that is bounded [0,1] (for Phi when 2x2 table). So they are "simply" normalized Chi-sq values.

I have seen around the web uses of Phi as an effect size for 1D goodness of fit tests, where it can take values larger than 1, but generally "larger = more deviation from the expected values" (0 no deviation). This doesn't seem that weird to me...
[Rosenberg (2010)](https://doi.org/10.1371/journal.pone.0010059) suggests a method for obtaining an effect size that cannot be larger than 1 (which I have expanded here to that 1xK table).

(I've also seen Phi used for xtabs > 2x2, but really I see no reason for this since V is readily accessible.)

So I think (breaking change @IndrajeetPatil):
- Phi for tables > 2x2 will fail.
- Phi for goodness of fit (regardless of expected being uniform or not) will use my generalization of Rosenberg (2010).

**Comment by mattansb** [comment]
Here is another example of the "normalize by max-Chisq" for goodness of fit https://doi.org/10.2466%2Fpms.103.2.412-414

**Comment by bwiernik** [comment]
That shouldn't be called phi though. Phi is explicitly defined as a correlation between two binary variables.

**Comment by mattansb** [comment]
Counter point: Cohen's *w* is defined as the normalized difference between expected and observed relative frequencies

![image](https://user-images.githubusercontent.com/35330040/165358365-941bf0d8-996b-400f-87ef-a5fe6e9bd07f.png)

Which is equivalent to phi, so...?

What would you call this effect size, if not an adjusted / normalized Phi for goodness-of-fit?

**Comment by bwiernik** [comment]
w or normalized chi-square or E_chi-square. Anything by phi

**Comment by mattansb** [comment]
Okay, here is what I have so far:

| Problem type | Name | Equation | Bounded [0,1] | Notes |
|------|-----|-----|------|-----|
| 2x2 table | Phi | `sqrt(Chisq/N)` | Yes | |
| >2x2 table | Crame'r V | `phi/sqrt(min(cols,rows)-1)` | Yes | |
| Goodness of fit ***OR*** any sized xtab | Cohen's w | `sqrt(Chisq/N)` | Only when 1x2 with uniform expected, or 2x2 xtab | Although equivilant to phi, Cohen's explicitly defines it for goodness of fit and any sized xtab |
| Goodness of fit | ***?????*** | `sqrt(Chisq/(N*(1-q)/q)` | Yes | |

Still looking for a name for the last one. 

Phi, V, and ???? can be unified under a single function, `xtab_correlation()`, perhaps, with the `phi()` and `cramers_v()` serving as aliases.

**Comment by mattansb** [comment]
I do *not* like the name of this index 😥😭

``` r
library(effectsize)

Smoking_ASD <- as.table(c(ASD = 17, ASP = 11, TD = 640))

(Xsq <- chisq.test(Smoking_ASD, p = c(0.015, 0.010, 0.975)))
#> 
#>  Chi-squared test for given probabilities
#> 
#> data:  Smoking_ASD
#> X-squared = 7.8521, df = 2, p-value = 0.01972

chisq_to_correlation(
  7.8521,
  n = sum(Smoking_ASD),
  nrow = 1,
  ncol = 3,
  p = c(0.015, 0.010, 0.975)
)
#> r    |       95% CI
#> -------------------
#> 0.01 | [0.00, 1.00]
#> 
#> - Adjusted for non-uniform expected probabilities.
#> - One-sided CIs: upper bound fixed at [1.00].

effectsize(Xsq)
#> r    |       95% CI
#> -------------------
#> 0.01 | [0.00, 1.00]
#> 
#> - Adjusted for non-uniform expected probabilities.
#> - One-sided CIs: upper bound fixed at [1.00].

chisq_correlation(Smoking_ASD, p = c(0.015, 0.010, 0.975))
#> r    |       95% CI
#> -------------------
#> 0.01 | [0.00, 1.00]
#> 
#> - Adjusted for non-uniform expected probabilities.
#> - One-sided CIs: upper bound fixed at [1.00].
```

<sup>Created on 2022-05-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by mattansb** [comment]
- Adjusted Cohen's w (`w (Adj.)` / Standardized Cohen's w (`w (Std.)`) (a-la Johnstom et al 2006)?
- Chi correlation (not squared!) - `r (Chi)`
- Standardized Chi - `Chi (Std.)`
- Deviation index? Badness of fit (BOF)? Anti correlation? (0 is perfect fit, 1 is worst)

**Comment by bwiernik** [comment]
I suggested "Normalized Chi2". It's normalized not standardized

**Comment by mattansb** [comment]
`normalized_chi()` (not chi2, because it's square-rooted?) with column name `Norm. Chi`?

I can live with that, I guess...

**Comment by mattansb** [comment]
Alright! Done!

``` r
library(effectsize)

## 2-by-2 tables
```

``` r
RCT <-
  matrix(c(71, 30,
           50, 100), nrow = 2, byrow = TRUE,
         dimnames = list(
           Diagnosis = c("Sick", "Recovered"),
           Group = c("Treatment", "Control")))
RCT # note groups are COLUMNS
#>            Group
#> Diagnosis   Treatment Control
#>   Sick             71      30
#>   Recovered        50     100

phi(RCT)
#> Phi  |       95% CI
#> -------------------
#> 0.36 | [0.26, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].
cohens_w(RCT) # same
#> Cohen's w |       95% CI
#> ------------------------
#> 0.36      | [0.26, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].
cramers_v(RCT) # same 
#> Cramer's V |       95% CI
#> -------------------------
#> 0.36       | [0.26, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].

pearsons_c(RCT)
#> Pearson's C |       95% CI
#> --------------------------
#> 0.34        | [0.25, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].

## Larger tables
```

``` r
M <-
  matrix(c(150, 100, 165,
           130, 50, 65,
           35, 10, 2,
           55, 40, 25), nrow = 4,
         dimnames = list(
           Music = c("Pop", "Rock", "Jazz", "Classic"),
           Study = c("Psych", "Econ", "Law")))
M
#>          Study
#> Music     Psych Econ Law
#>   Pop       150   50   2
#>   Rock      100   65  55
#>   Jazz      165   35  40
#>   Classic   130   10  25

cohens_w(M)
#> Cohen's w |       95% CI
#> ------------------------
#> 0.34      | [0.27, 1.41]
#> 
#> - One-sided CIs: upper bound fixed at [1.41~].

cramers_v(M)
#> Cramer's V |       95% CI
#> -------------------------
#> 0.24       | [0.19, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].

pearsons_c(M)
#> Pearson's C |       95% CI
#> --------------------------
#> 0.32        | [0.26, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].


## Goodness of fit
```

``` r
Smoking_ASD <- as.table(c(ASD = 17, ASP = 11, TD = 640))

normalized_chi(Smoking_ASD)
#> Norm. Chi |       95% CI
#> ------------------------
#> 0.94      | [0.89, 1.00]
#> 
#> - Adjusted for uniform expected probabilities.
#> - One-sided CIs: upper bound fixed at [1.00].

cohens_w(Smoking_ASD)
#> Cohen's w |      95% CI
#> -----------------------
#> 1.33      | [1.26, Inf]
#> 
#> - One-sided CIs: upper bound fixed at [Inf].

pearsons_c(Smoking_ASD)
#> Pearson's C |       95% CI
#> --------------------------
#> 0.80        | [0.78, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].

# Use custom expected values:
normalized_chi(Smoking_ASD, p = c(0.015, 0.010, 0.975))
#> Norm. Chi |       95% CI
#> ------------------------
#> 0.01      | [0.00, 1.00]
#> 
#> - Adjusted for non-uniform expected probabilities.
#> - One-sided CIs: upper bound fixed at [1.00].

cohens_w(Smoking_ASD, p = c(0.015, 0.010, 0.975))
#> Cohen's w |      95% CI
#> -----------------------
#> 0.11      | [0.03, Inf]
#> 
#> - One-sided CIs: upper bound fixed at [Inf].

pearsons_c(Smoking_ASD, p = c(0.015, 0.010, 0.975))
#> Pearson's C |       95% CI
#> --------------------------
#> 0.11        | [0.03, 1.00]
#> 
#> - One-sided CIs: upper bound fixed at [1.00].
```

<sup>Created on 2022-05-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
